Automattic/wp-calypso
# Draft: Add site language switcher to gutenboarding

Build on top of the work that @sirreal did in #39627 this PR allows live switching of the language in gutenboarding, and adds a language picker made from core components.

I changed from using an actual modal component to using a step because that is now how the domains and plans steps have been implemented.

I've got it behind a feature flag for now because I'd like to get the language switching changes and refactoring tested and merged before the UI is completely polished.

### Functionality
Can now change the site language dynamically in gutenboarding using the `window.updateLocale('en')` function

### Changes
* Wraps the app in a new LocaleContext component which exposes a function for changing the locale.
* Breaks the initial locale loading code out of gutenboarding/index.tsx and into `locale-context.tsx`.
* Handles chunked translations as well as non chunked translations
* Includes the actual language picker now as a modal step
* The language picker doesn't currently have search functionality, I'd like to implement that as an additional PR because this one is already big! 

### Limitations
* Does not set the language slug in the URL, language gets lost on refresh
* Does not set the language in the users settings after completing the flow
* Search for languages is not implemented

Part of https://github.com/Automattic/wp-calypso/issues/40011

### Testing instructions

* This is hidden behind a feature flag, enable it through the URL by appending `?flags=gutenboarding/language-picker`
* The <HTML> Element should have a `lang` attribute set appropriately for the current language
* The html element should have its `dir` attribute set to `rtl` for right to left languages such as arabic
* The language picker step should behave similarly to the plans and domain picker steps
* In production, we use Chunked Translations which is a different way of loading translations, affected by this PR. 
    * Chunked translations can be tested on ci by adding `?flags=gutenboarding/language-picker,` 
    * Or tested locally by running the following command `ENTRY_LIMIT=entry-gutenboarding BUILD_TRANSLATION_CHUNKS=true ENABLE_FEATURES=use-translation-chunks,gutenboarding/language-picker yarn start`

<img width="1187" alt="Screen Shot 2020-07-14 at 7 16 46 PM" src="https://user-images.githubusercontent.com/22446385/87409249-0be7cc80-c608-11ea-9157-0f9245a41eaf.png">

fixes #40011
